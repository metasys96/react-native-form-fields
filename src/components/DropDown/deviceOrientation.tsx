import { useState, useEffect } from "react";
import { Dimensions } from "react-native";

function DeviceOrientation() {
  const [deviceOrientation, setDeviceOrientation] = useState('portrait');

  useEffect(() => {
    function updateState() {
      const { height, width } = Dimensions.get("window");
      if (height >= width) {
        setDeviceOrientation('portrait');
      } else {
        setDeviceOrientation('landscape');
      }
    }

    updateState();
    Dimensions.addEventListener("change", updateState);
    return () => Dimensions.removeEventListener("change", updateState);
  }, []);

  return deviceOrientation;
}

export default DeviceOrientation;
