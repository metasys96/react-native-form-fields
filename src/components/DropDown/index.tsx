import React, {useState, useEffect} from 'react';
import {StyleSheet} from 'react-native';
import {
  Modal,
  Picker,
  Platform,
  Text,
  TextInput,
  TouchableOpacity,
  TouchableWithoutFeedback,
  View,
} from 'react-native';
import DeviceOrientation from './deviceOrientation';
import isEqual from 'lodash.isequal';
import {Icon} from 'react-native-elements';

interface AppProps {
  items: Array<itemsObj>;
  useNativeAndroidPickerStyle?: boolean;
  style?:any;
  modalProps?: any;
  dropDownProps?: any;
  disabled?: boolean;
  textInputProps?: any;
  placeholderTextColor?: string;
  placeholder?: dropDownplaceHolder;
  selectText?: string;
  hideSelectBar?: boolean;
  onChange?: (value: string , index: number)=>void;
  selectedItem?: any;
  dropdownIcon?: dropDownIcon;
  fieldIcon?: boolean;
  itemTextColor?: string;
}

const DropDown = (props: AppProps) => {
  const {
    useNativeAndroidPickerStyle,
    style,
    modalProps,
    dropDownProps,
    disabled,
    textInputProps,
    placeholderTextColor,
    placeholder,
    selectText,
    hideSelectBar,
    onChange,
    dropdownIcon,
    fieldIcon,
    itemTextColor,
  } = props;

  const handlePlaceholder = (placeholderhere: any) => {
    if (props.items) {
      if (isEqual(placeholderhere, {}) || isEqual(placeholderhere, undefined)) {
        return props.items;
      } else {
        return [placeholderhere, ...props.items];
      }
    }
  };

  const merged = handlePlaceholder(placeholder);
  const [items] = useState(merged);
  const [selectedItem, setSelectedItem] = useState(
    props.selectedItem ? props.selectedItem : placeholder,
  );
  const [showPicker, setShowPicker] = useState(false);
  const deviceOrientation = DeviceOrientation();
  const [orientation] = useState(deviceOrientation);

  const defaultStyles = StyleSheet.create({
    viewContainer: {
      alignSelf: 'stretch',
    },
    iconContainer: {
      position: 'absolute',
      right: 0,
    },
    modalViewTop: {
      flex: 1,
    },
    modalViewMiddle: {
      height: 44,
      flexDirection: 'row',
      justifyContent: 'flex-end',
      alignItems: 'center',
      paddingHorizontal: 10,
      backgroundColor: '#f5f5f5',
      borderTopWidth: 0.5,
      borderTopColor: '#919498',
      zIndex: 2,
    },
    chevronContainer: {
      flexDirection: 'row',
    },
    chevron: {
      width: 15,
      height: 15,
      backgroundColor: 'transparent',
      borderColor: '#D0D4DB',
      borderTopWidth: 1.5,
      borderRightWidth: 1.5,
    },
    chevronUp: {
      marginLeft: 11,
      transform: [{translateY: 4}, {rotate: '-45deg'}],
    },
    chevronDown: {
      marginLeft: 22,
      transform: [{translateY: -5}, {rotate: '135deg'}],
    },
    chevronActive: {
      borderColor: '#007AFE',
    },
    select: {
      color: '#007AFE',
      fontWeight: 'bold',
      fontSize: 15,
      paddingTop: 1,
      paddingRight: 2,
    },
    modalViewBottom: {
      justifyContent: 'center',
      backgroundColor: '#F9f9f9',
    },
    placeholder: {
      color: '#C7C7CD',
    },
    headlessAndroidPicker: {
      position: 'absolute',
      width: '100%',
      height: '100%',
      color: 'transparent',
      opacity: 0,
    },
    selectedItem: {color: 'black'},
  });

  useEffect(() => {
    items &&
      items.forEach((element: any, index: number) => {
        if (element && element.value === selectedItem.value) {
          onChange(selectedItem.value, index);
        }
      });
  }, [items]);

  const onValueChanges = (value: string, index: number) => {
    setSelectedItem(items[index]);
    onChange(value, index);
  };

  const getPlaceholderStyle = () => {
    if (
      !isEqual(placeholder, {}) &&
      placeholder &&
      selectedItem.label === placeholder.label
    ) {
      return {
        ...defaultStyles.placeholder,
        ...style.placeholder,
        color: placeholderTextColor,
      };
    } else if (isEqual(placeholder, {})) {
      return {
        ...defaultStyles.selectedItem,
        ...style.selectedItem,
      };
    } else {
      return {};
    }
  };

  const renderPickerItemList = () => {
    return (
      items &&
      items.map((item: any) => {
        if (item) {
          return (
            <Picker.Item
              label={item.label ? item.label : ''}
              value={item.value ? item.value : ''}
              key={item.key ? item.key : item.label}
              color={item.color ? item.color : itemTextColor}
            />
          );
        }
      })
    );
  };

  const renderPickerBar = () => {
    if (hideSelectBar) {
      return null;
    }
    return (
      <View style={[defaultStyles.modalViewMiddle, style.modalViewMiddle]}>
        <TouchableWithoutFeedback
          onPress={() => setShowPicker(!showPicker)}
          hitSlop={{top: 4, right: 4, bottom: 4, left: 4}}>
          <View>
            <Text
              allowFontScaling={false}
              style={[defaultStyles.select, style.select]}>
              {selectText}
            </Text>
          </View>
        </TouchableWithoutFeedback>
      </View>
    );
  };

  const renderIcon = () => {
    return (
      <View
        style={[
          {marginRight: 5},
          defaultStyles.iconContainer,
          style.iconContainer,
        ]}>
        {fieldIcon ? (
          fieldIcon
        ) : (
          <Icon
            size={dropdownIcon.size}
            name={dropdownIcon.name}
            type={dropdownIcon.type}
            color={dropdownIcon.color}
          />
        )}
      </View>
    );
  };

  const renderTextInput = () => {
    const containerStyle =
      Platform.OS === 'ios'
        ? style.inputIOSContainer
        : style.inputAndroidContainer;
    if (selectedItem) {
      return (
        <View
          pointerEvents="box-only"
          style={[DDLStyles.placeHolder, containerStyle]}>
          <TextInput
            style={[
              Platform.OS === 'ios' ? style.inputIOS : style.inputAndroid,
              getPlaceholderStyle(),
            ]}
            value={selectedItem.value ? selectedItem.value : selectedItem.label}
            editable={false}
            {...textInputProps}
          />
          {renderIcon()}
        </View>
      );
    }
  };

  const renderIOS = () => {
    const animationType =
      modalProps && modalProps.animationType
        ? modalProps.animationType
        : 'slide';

    return (
      <View style={[defaultStyles.viewContainer, style.viewContainer]}>
        <TouchableWithoutFeedback onPress={() => setShowPicker(!showPicker)}>
          {renderTextInput()}
        </TouchableWithoutFeedback>

        <Modal
          visible={showPicker}
          transparent
          animationType={animationType}
          supportedOrientations={['portrait', 'landscape']}
          {...modalProps}>
          <TouchableOpacity
            style={[defaultStyles.modalViewTop, style.modalViewTop]}
            onPress={() => setShowPicker(!showPicker)}
          />
          {renderPickerBar()}
          <View
            style={[
              defaultStyles.modalViewBottom,
              {height: orientation === 'portrait' ? 215 : 162},
              style.modalViewBottom,
            ]}>
            <Picker
              enabled={!disabled}
              onValueChange={(value, index) => onValueChanges(value, index)}
              selectedValue={
                selectedItem && selectedItem.value
                  ? selectedItem.value
                  : items && items[0].value
                  ? items[0].value
                  : null
              }
              {...dropDownProps}>
              {renderPickerItemList()}
            </Picker>
          </View>
        </Modal>
      </View>
    );
  };

  const renderAndroidHeadless = () => {
    return (
      <TouchableWithoutFeedback onPress={() => setShowPicker(!showPicker)}>
        <View style={[style.headlessAndroidContainer]}>
          {renderTextInput()}
          <Picker
            style={[
              dropdownIcon ? DDLStyles.transparent : {},
              defaultStyles.headlessAndroidPicker,
              style.headlessAndroidPicker,
            ]}
            enabled={!disabled}
            onValueChange={(value: string, index: number) =>
              onValueChanges(value, index)
            }
            selectedValue={selectedItem.value}
            {...dropDownProps}>
            {renderPickerItemList()}
          </Picker>
        </View>
      </TouchableWithoutFeedback>
    );
  };

  const renderAndroidNativePickerStyle = () => {
    return (
      <View style={[defaultStyles.viewContainer, style.viewContainer]}>
        <Picker
          style={[
            dropdownIcon ? DDLStyles.transparent : {},
            style.inputAndroid,
            getPlaceholderStyle(),
          ]}
          enabled={!disabled}
          onValueChange={(value: string, index: number) =>
            onValueChanges(value, index)
          }
          selectedValue={selectedItem.value}
          {...dropDownProps}>
          {renderPickerItemList()}
        </Picker>
        {renderIcon()}
      </View>
    );
  };

  return Platform.OS === 'ios'
    ? renderIOS()
    : useNativeAndroidPickerStyle === true
    ? renderAndroidNativePickerStyle()
    : renderAndroidHeadless();
};

DropDown.defaultProps = {
  placeholder: {
    id: 'default',
    label: 'Please select',
    value: '',
    color: '#9EA0A4',
  },
  disabled: false,
  style: {},
  placeholderTextColor: '#C7C7CD',
  itemTextColor: 'black',
  useNativeAndroidPickerStyle: false,
  hideSelectBar: false,
  selectText: 'Select',
  dropdownIcon: {
    size: 20,
    name: 'angle-down',
    type: 'font-awesome',
    color: 'black',
  },
  modalProps: {},
  textInputProps: {},
  dropDownProps: {},
};

const DDLStyles = StyleSheet.create({
  transparent: {backgroundColor: 'transparent'},
  placeHolder: {
    width: '100%',
    borderWidth: 1,
    borderColor: 'lightgrey',
    height: 40,
    justifyContent: 'center',
    paddingLeft: 5,
    paddingRight: 5,
    color: 'black',
  },
});

export default DropDown;
