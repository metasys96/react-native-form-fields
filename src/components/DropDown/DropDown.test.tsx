import {shallow} from 'enzyme';
import DropDown from './index';
import React from 'react';
import 'react-native';
import {cleanup, fireEvent} from 'react-native-testing-library';
import renderer from 'react-test-renderer';
import {Picker} from 'react-native';

afterEach(cleanup);

const dropDownOptions = [
  {label: 'RD', value: 'Red'},
  {label: 'GR', value: 'Green'},
  {label: 'BL', value: 'Blue'},
];

function getShallowComponent() {
  return shallow(
    <DropDown
      items={dropDownOptions}
      selectedItem={{label: 'BL', value: 'Blue'}}
    />,
  );
}
function getJSONRenderer() {
  return renderer.create(<DropDown />).toJSON();
}

test('create snapshot', () => {
  const component = getJSONRenderer();
  expect(component).toMatchSnapshot();
});

it('renders component', () => {
  const component = getJSONRenderer();
  expect(component).toBeTruthy();
});

it('should contain Picker and text as same as length of the options', () => {
  const component = getShallowComponent();
  expect(component.find('Picker')).toHaveLength(1);
  expect(component.find('Text').length).toEqual(1);
});

it('should change the dropdown option on press function', async () => {
  const onPressFunction = jest.fn();
  const wrapper = shallow(
    <DropDown
      items={dropDownOptions}
      selectedItem={{label: 'BL', value: 'Blue'}}
      onChange={onPressFunction}
    />,
  );

  const valueChangeEvent = {target: {value: 'Red', index: 1}};
  wrapper
    .find(Picker)
    .first()
    .simulate('onChange', valueChangeEvent);

  expect(onPressFunction).toHaveBeenCalledTimes(1);

});
