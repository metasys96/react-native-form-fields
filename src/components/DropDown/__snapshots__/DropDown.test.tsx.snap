
exports[`create snapshot 1`] = `
<View
  style={
    Array [
      Object {
        "alignSelf": "stretch",
      },
      undefined,
    ]
  }
>
  <View
    accessible={true}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    pointerEvents="box-only"
    style={
      Array [
        Object {
          "borderColor": "lightgrey",
          "borderWidth": 1,
          "height": 30,
          "justifyContent": "center",
          "paddingLeft": 5,
          "paddingRight": 5,
          "width": "100%",
        },
        undefined,
      ]
    }
  >
    <TextInput
      allowFontScaling={true}
      editable={false}
      rejectResponderTermination={true}
      style={
        Array [
          undefined,
          Object {
            "color": "#C7C7CD",
          },
        ]
      }
      underlineColorAndroid="transparent"
      value="Please select"
    />
    <View
      style={
        Array [
          Object {
            "marginRight": 5,
          },
          Object {
            "position": "absolute",
            "right": 0,
          },
          undefined,
        ]
      }
    >
      <View>
        <View
          style={
            Object {
              "alignItems": "center",
              "backgroundColor": "transparent",
              "justifyContent": "center",
            }
          }
          theme={
            Object {
              "colors": Object {
                "disabled": "hsl(208, 8%, 90%)",
                "divider": "#bcbbc1",
                "error": "#ff190c",
                "grey0": "#393e42",
                "grey1": "#43484d",
                "grey2": "#5e6977",
                "grey3": "#86939e",
                "grey4": "#bdc6cf",
                "grey5": "#e1e8ee",
                "greyOutline": "#bbb",
                "platform": Object {
                  "android": Object {
                    "error": "#f44336",
                    "primary": "#2196f3",
                    "secondary": "#9C27B0",
                    "success": "#4caf50",
                    "warning": "#ffeb3b",
                  },
                  "ios": Object {
                    "error": "#ff3b30",
                    "primary": "#007aff",
                    "secondary": "#5856d6",
                    "success": "#4cd964",
                    "warning": "#ffcc00",
                  },
                },
                "primary": "#2089dc",
                "searchBg": "#303337",
                "secondary": "#8F0CE8",
                "success": "#52c41a",
                "warning": "#faad14",
              },
            }
          }
          underlayColor="white"
        >
          <Text
            allowFontScaling={false}
            style={
              Array [
                Object {
                  "color": "black",
                  "fontSize": 20,
                },
                Object {
                  "backgroundColor": "transparent",
                },
                Object {
                  "fontFamily": "FontAwesome",
                  "fontStyle": "normal",
                  "fontWeight": "normal",
                },
                Object {},
              ]
            }
            testID="iconIcon"
          >
            ï„‡
          </Text>
        </View>
      </View>
    </View>
  </View>
  <Modal
    animationType="slide"
    hardwareAccelerated={false}
    supportedOrientations={
      Array [
        "portrait",
        "landscape",
      ]
    }
    transparent={true}
    visible={false}
  >
    <View
      accessible={true}
      focusable={true}
      isTVSelectable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "flex": 1,
          "opacity": 1,
        }
      }
    />
    <View
      style={
        Array [
          Object {
            "alignItems": "center",
            "backgroundColor": "#f5f5f5",
            "borderTopColor": "#919498",
            "borderTopWidth": 0.5,
            "flexDirection": "row",
            "height": 44,
            "justifyContent": "flex-end",
            "paddingHorizontal": 10,
            "zIndex": 2,
          },
          undefined,
        ]
      }
    >
      <View
        accessible={true}
        focusable={true}
        hitSlop={
          Object {
            "bottom": 4,
            "left": 4,
            "right": 4,
            "top": 4,
          }
        }
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
      >
        <Text
          allowFontScaling={false}
          style={
            Array [
              Object {
                "color": "#007AFE",
                "fontSize": 15,
                "fontWeight": "bold",
                "paddingRight": 2,
                "paddingTop": 1,
              },
              undefined,
            ]
          }
        >
          Select
        </Text>
      </View>
    </View>
    <View
      style={
        Array [
          Object {
            "backgroundColor": "#F9f9f9",
            "justifyContent": "center",
          },
          Object {
            "height": 215,
          },
          undefined,
        ]
      }
    >
      <View>
        <RCTPicker
          items={Array []}
          onChange={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          selectedIndex={0}
          style={
            Array [
              Object {
                "height": 216,
              },
              undefined,
            ]
          }
        />
      </View>
    </View>
  </Modal>
</View>
`;
